src/box/txn_limbo.h

```diff
@@ -90,14 +85,15 @@ txn_limbo_entry_is_complete(const struct txn_limbo_entry *e)
  *     as they wouldn't depend on each other directly.
  */
 struct txn_limbo {
+	struct txn_limbo_entry head_entry;
 	/**
 	 * Queue of limbo entries. Ordered by LSN. Some of the
 	 * entries in the end may not have an LSN yet (their local
 	 * WAL write is still in progress), but their order won't
 	 * change anyway. Because WAL write completions will give
 	 * them LSNs in the same order.
 	 */
-	struct rlist queue;
+	struct rlist *queue;
```

Comment:
I would suggest to reconsider the storage approach here. Lets keep the queue as is. What is wrong with that? Why do you need the queue to be a pointer?

Reply:
I decided to add a dummy entry `head_entry` which is the head of the list. This significantly simplifies the code and allows you to avoid several checks.
https://github.com/Astronomax/tarantool/blob/6eee079a537a40c9721f6ca8a9fe616748f3b5c6/src/box/txn_limbo.h#L88-L92
Therefore `queue` is forced to be a pointer.

Reply:
From what I see, because of this check you had to introduce the flag, and check `!rlist_is_head` in many places. Seems like not really better than checking for `==/!= NULL`, hm? Would you please try to delete the dummy entry and respond here if you find any specifically ugly places?

Reply:
This avoided several checks due to the fact that `head_entry` has lsn = -1. But that was when the `txn_limbo_update_confirmed_entry_after_rollback` function was. Now this does not matter, there will be the same number of checks, you’re right.

Reply:
Lets then drop the dummy entry? (Sorry, didn't see the new patch version yet, just looking at comments!)